File: aoc-2023/cpp/day4.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

int find_matches(std::string card) {
  int sum = 0;
  std::string numbers = card.substr(card.find(':') + 2);
  std::unordered_map<int, int> map;
  bool found_divisor = false;

  std::string number_string = "";
  for (int i = 0; i < numbers.size(); i++) {
    if (!std::isdigit(numbers[i]) && number_string.size() > 0) {
      if (!found_divisor) {
        map[std::stoi(number_string)]++;
      } else if (found_divisor) {
        if (map.find(std::stoi(number_string)) != map.end()) {
          sum++;
        }
      }
      number_string = "";
    } else if (numbers[i] == '|') {
      found_divisor = true;
      continue;
    } else if (std::isdigit(numbers[i])) {
      number_string += numbers[i];
    }
  }

  if (map.find(std::stoi(number_string)) != map.end()) {
    sum++;
  }

  return sum;
}

int part_one(std::string card) {
  int matches = find_matches(card);
  return matches > 1 ? 1 << (matches - 1) : matches;
}

int part_two(std::vector<std::string> *cards) {
  std::unordered_map<int, int> map;
  int current_card_id = 0;
  int total_cards = 0;

  for (int i = 0; i < cards->size(); i++) {
    current_card_id = i + 1;
    std::string card = cards->at(i);
    int matches = find_matches(card);
    for (int j = 1; j <= matches; j++) {
      int copies = map[current_card_id];
      map[current_card_id + j] += 1 + copies;
    }
  }

  for (const auto c : map) {
    total_cards += c.second;
  }

  return total_cards + current_card_id;
}

int main(int argc, char *argv[]) {
  std::fstream file(argv[1]);
  std::string line;
  std::vector<std::string> cards;

  int part_one_total = 0;
  int part_two_total = 0;

  while (std::getline(file, line)) {
    part_one_total += part_one(line);
    cards.push_back(line);
  }

  part_two_total = part_two(&cards);

  std::cout << "Part 1: " << part_one_total << "\n";
  std::cout << "Part 2: " << part_two_total << "\n";
}

File: aoc-2023/src/day4.rs
use std::collections::HashMap;

use crate::timed;

fn find_matches(line: &str) -> i32 {
    let mut sum = 0;
    let mut found_divisor: bool = false;
    let mut map: HashMap<i32, i32> = HashMap::new();

    let numbers = &line[line.find(':').unwrap()..];

    let mut number_string = String::new();

    for c in numbers.chars() {
        if !c.is_ascii_digit() && !number_string.is_empty() {
            if !found_divisor {
                let n = number_string.parse::<i32>().unwrap();
                map.insert(n, 1);
            } else if found_divisor && map.contains_key(&number_string.parse::<i32>().unwrap()) {
                sum += 1;
            }
            number_string = String::new();
        } else if c == '|' {
            found_divisor = true;
            continue;
        } else if c.is_ascii_digit() {
            number_string.push(c);
        }
    }

    if map.contains_key(&number_string.parse::<i32>().unwrap()) {
        sum += 1;
    }

    sum
}

fn part_one(card: &str) -> i32 {
    let matches = find_matches(card);
    if matches > 1 {
        1 << (matches - 1)
    } else {
        matches
    }
}

fn part_two(cards: &str) -> i32 {
    let mut map: HashMap<i32, i32> = HashMap::new();
    let mut current_card_id = 0;
    let mut total_cards = 0;

    for (i, card) in cards.lines().enumerate() {
        current_card_id = i as i32 + 1;
        let matches = find_matches(card);
        for j in 1..=matches {
            let copies = *map.get(&current_card_id).unwrap_or(&0);
            if let std::collections::hash_map::Entry::Vacant(e) = map.entry(current_card_id + j) {
                e.insert(1 + copies);
            } else {
                *map.get_mut(&(current_card_id + j)).unwrap() += 1 + copies;
            }
        }
    }

    for &copies in map.values() {
        total_cards += copies;
    }

    total_cards + current_card_id
}

pub fn run(part: Option<u8>) {
    let input = include_str!("../../inputs/2023/day4/input.txt");

    match part.unwrap_or_default() {
        1 => timed(part_one, input, 1),
        2 => timed(part_two, input, 2),
        _ => {
            timed(part_one, input, 1);
            timed(part_two, input, 2)
        }
    };
}
